Default to ES modules with single Rollup config

colinrotherham opened the pull request:
This PR consolidates all [`govuk-frontend` npm package](https://www.npmjs.com/package/govuk-frontend) JavaScript into the `govuk` directory

We use Rollup to build JavaScript into the following formats + extensions:

* ECMAScript (ES) modules as `*.mjs`
* ECMAScript (ES) module bundles as `*.bundle.mjs`
* Universal Module Definition (UMD) bundles as `*.bundle.js`

But like we do for our GitHub release zip files also add a "ready to go" minified bundle for:

* ECMAScript (ES) module bundle minified as `govuk-frontend.min.js`

### CommonJS compatibility

We've felt the pain of ["ESM only" packages](https://github.com/alphagov/govuk-frontend/commit/94951716e65702594ecd5acd9b55b71b8a19b93d) recently and know that CommonJS compatibility is important for both [Node.js `require('govuk-frontend')` package resolution](https://github.com/alphagov/govuk-frontend/issues/3755) and for test runners that don't support ES modules yet

By dropping CommonJS and going "ESM only" we'd hit issues such as:

1. [Jest ECMAScript modules support](https://jestjs.io/docs/ecmascript-modules) needing `--experimental-vm-modules`
2. [Mocha current limitations](https://mochajs.org/#current-limitations) like [`--watch` mode](https://github.com/mochajs/mocha/issues/4374) only working with CommonJS

Although more complex workarounds like [Babel compiling ES modules to CommonJS](https://github.com/alphagov/govuk-frontend/commit/b6e8efba6007fb6f2e29230e61a9d829a1660c09) can work instead

### Legacy bundlers

Some bundlers like [Browserify support `require('govuk-frontend')`](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#import-javascript-using-a-bundler) but may need AMD not CommonJS format. Similarly the [unpkg CDN](https://unpkg.com) falls back to expect a [UMD bundle in **package.json** `"main"`](https://unpkg.com/#examples)

### Package layout

```shell
dist
└── govuk
    ├── govuk-frontend.min.js         # ECMAScript (ES) module bundle, minified
    │
    ├── all.mjs                       # ECMAScript (ES) module
    ├── all.bundle.mjs                # ECMAScript (ES) module bundle
    ├── all.bundle.js                 # Universal Module Definition (UMD) bundle
    │
    └── components
        ├── accordion
        │   ├── accordion.mjs         # ECMAScript (ES) module
        │   ├── accordion.bundle.mjs  # ECMAScript (ES) module bundle
        │   └── accordion.bundle.js   # Universal Module Definition (UMD) bundle
        │
        └── button
            ├── button.mjs            # ECMAScript (ES) module
            ├── button.bundle.mjs     # ECMAScript (ES) module bundle
            └── button.bundle.js      # Universal Module Definition (UMD) bundle
```

### Script includes

But by adding "ready to go" bundles `all.bundle.js` (UMD) and `all.bundle.mjs` (ES module) we maintain non-bundler compatibility even when importing JavaScript directly ([see **README.md**](https://github.com/alphagov/govuk-frontend/blob/main/packages/govuk-frontend/README.md#importing-javascript))

✅ Rails Asset Pipeline using `//= require` to concatenate UMD `all.bundle.js`
✅ Browsers using dynamic `await import()` to load ES module `all.bundle.mjs`
✅ Browsers using `<script type="module">` to load ES module `all.bundle.mjs`

```html
<script src="/javascripts/all.bundle.js"></script>
<script>window.GOVUKFrontend.initAll()</script>
```

```html
<script type="module">
  import { initAll } from '/javascripts/all.bundle.mjs'
  initAll()
</script>
```

**Note:** Only `govuk-frontend.min.js` is minified, unlike other bundles

### Window globals

~What do we do about UMD-only window globals like `window.GOVUKFrontend.Accordion`?~

Regarding window globals we'll continue to support UMD bundles in v5 but will deprecate them in a future release

colinrotherham commented:
Thanks @romaricpascal

Yeah it's primarily to address this ["modules versus bundles" thread](https://github.com/alphagov/govuk-frontend/pull/3386#discussion_r1155949622) which caught us out

Historically we've bundled every JavaScript file separately into UMD, but now we can pick any module format and even CommonJS bundlers should be able to do tree-shaking

colinrotherham commented:
@romaricpascal I've moved CommonJS to `*.js` (not `*.cjs`)  and pushed again

Regarding the **package.json** `"default"` field the [Node.js docs for **Conditional exports**](https://nodejs.org/api/packages.html#conditional-exports) says:

>When using environment branches, always include a "default" condition where possible. Providing a "default" condition ensures that any unknown JS environments are able to use this universal implementation, which helps avoid these JS environments from having to pretend to be existing environments in order to support packages with conditional exports. For this reason, using "node" and "default" condition branches is usually preferable to using "node" and "browser" condition branches.

Yet we need `"browser"` to avoid Browserify + webpack v4 confusion over bundles. Also Rollup [`@rollup/plugin-commonjs`](https://www.npmjs.com/package/@rollup/plugin-commonjs) won't understand `"node"` so I think we have a sensible compromise here 🤷‍♂️

colinrotherham commented:
Ready for review but we're going to write up a decision document based on this PR

colinrotherham commented on packages/govuk-frontend-review/src/app.mjs:
```diff
@@ -9,6 +11,9 @@ import * as middleware from './common/middleware/index.mjs'
 import * as routes from './routes/index.mjs'
 
+const require = createRequire(import.meta.url)
+const { version } = require('govuk-frontend/package.json')
```
@romaricpascal As discussed, we're using the local `govuk-frontend/package.json` not the shared config. This means the review app can still run `govuk-frontend@4` if installed into the workspace

I've had to use `require()` since ES modules get an experimental warning when running:

```mjs
import pkg from 'govuk-frontend/package.json' assert { type: 'json' }
```

Import assertions were reconsidered as [import attributes](https://github.com/tc39/proposal-import-attributes#readme) due to [security concerns](https://github.com/WICG/webcomponents/issues/839)

36degrees reviewed (commented):
This is looking pretty good to me @colinrotherham – thanks for breaking it down into sensible commits, that made it pretty easy to review despite all the changes.

There are a couple of minor code style things I picked up on that I think might be worth discussing.

In terms of bigger questions one of the things we're trying to do in this release is make it clearer what is part of the 'public API' and what is private. We're currently allowing access to everything by including `/govuk/*` in our exports, but should consider only including the public bits of our API instead, as suggested by [the node docs on package entry points](https://nodejs.org/api/packages.html#package-entry-points:~:text=The%20%22exports%22,for%20their%20package.)? I don't think we need to address that in this PR, but we may want to write up an issue to revisit it?

Separately – and I think this was my suggestion originally – but now that we've spent some time trying it out do we think it might be a mistake to include the version number in the filename? It feels like we're having to jump through quite a few hoops in the last commit to make it work in the review app, and I wonder if we should be trying to simplify it both for ourselves and our users. Appreciate this means revisiting https://github.com/alphagov/govuk-design-system-architecture/pull/28!

colinrotherham commented on shared/tasks/scripts.mjs:
```diff
@@ -56,28 +53,47 @@ export async function compileJavaScript ([modulePath, { configPath, srcPath, des
+        }
+
+        // Update basename with new extension
+        file.base = `${output.preserveModules ? file.name : `${file.name}.bundle`}${file.ext}`
```
Yeah, you can, but it makes you do daft things 🙃

Prefer your suggestions so I'll swap one in

colinrotherham replied:
Gone with this one and updated the comment to match:

```mjs
// Update basename with new extension and optional suffix
file.base = `${file.name}${output.preserveModules ? '' : '.bundle'}${file.ext}`
```

colinrotherham commented on .github/workflows/tests.yml:
```diff
@@ -228,17 +228,16 @@ jobs:
         conditions:
-          - require
-          - import
+          - [require]
```
Yeah fair shout, we can save this for another day

For context, we spiked **package.json** `"browser"` exports pre-bundled but otherwise separate modules:

```json
".": {
  "browser": {
    "sass": "./dist/govuk/all.scss",
    "import": "./dist/govuk/all.bundle.mjs",
    "require": "./dist/govuk/all.bundle.js",
    "default": "./dist/govuk/all.bundle.js"
  },
  "sass": "./dist/govuk/all.scss",
  "import": "./dist/govuk/all.mjs",
  "require": "./dist/govuk/all.cjs",
  "default": "./dist/govuk/all.bundle.js"
}
```

Would have needed those CommonJS exports though so we dropped it

```suggestion
          - [browser, require]
```

colinrotherham replied:
Split out into separate spike PR:

* https://github.com/alphagov/govuk-frontend/pull/3920

colinrotherham commented on packages/govuk-frontend/tasks/build/package.test.mjs:
```diff
@@ -120,31 +126,74 @@ describe('packages/govuk-frontend/dist/', () => {
+
+        // Look for ES import for each component
+        expect(contents).toContain(outdent`
+          import { Accordion } from './components/accordion/accordion.mjs';
+          import { Button } from './components/button/button.mjs';
+          import { CharacterCount } from './components/character-count/character-count.mjs';
+          import { Checkboxes } from './components/checkboxes/checkboxes.mjs';
+          import { ErrorSummary } from './components/error-summary/error-summary.mjs';
+          import { Header } from './components/header/header.mjs';
+          import { NotificationBanner } from './components/notification-banner/notification-banner.mjs';
+          import { Radios } from './components/radios/radios.mjs';
+          import { SkipLink } from './components/skip-link/skip-link.mjs';
+          import { Tabs } from './components/tabs/tabs.mjs';
+        `)
+
+        // Look for ES modules named exports
+        expect(contents).toContain('export { Accordion, Button, CharacterCount, Checkboxes, ErrorSummary, Header, NotificationBanner, Radios, SkipLink, Tabs, initAll };')
```
Yeah we could do, is it a blocker?

Happy to add a separate PR and we can fix [**govuk/components/globals.test.js**](https://github.com/alphagov/govuk-frontend/blob/main/packages/govuk-frontend/src/govuk/components/globals.test.js#L33) too

36degrees replied:
Happy to do it in a follow up PR

36degrees commented:
> > In terms of bigger questions one of the things we're trying to do in this release is make it clearer what is part of the 'public API' and what is private. We're currently allowing access to everything by including `/govuk/*` in our exports, but should consider only including the public bits of our API instead, as suggested by [the node docs on package entry points](https://nodejs.org/api/packages.html#package-entry-points:~:text=The%20%22exports%22,for%20their%20package.)? I don't think we need to address that in this PR, but we may want to write up an issue to revisit it?
> 
> Yeah seems totally sensible
> 
> For anyone else reading this, I'd suggest we run:
> 
> ```shell
> npm run build
> npm pack --workspace govuk-frontend
> ```
> 
> You'll see **govuk-frontend-4.6.0.tgz** which:
> 
> 1. Includes all the paths listed in [**package.json** `"files"`](https://github.com/alphagov/govuk-frontend/blob/main/packages/govuk-frontend/package.json#L8)
> 2. Excludes all the paths listed in [**.npmignore**](https://github.com/alphagov/govuk-frontend/blob/main/packages/govuk-frontend/.npmignore)
> 
> Would love your thoughts on what to exclude (even internal-but-not-private things like **I18n**)

To avoid blocking this PR any further, shall we create an issue to look at which files/globs we include in export? (Before we ship v5)

> Don't worry, I remember we also wanted to address all those files named `all.*` in every flavour 😮
> 
> What would be your suggested filename?

Just remove the version number, leaving `govuk-frontend.min.js`?

I don't actually remember what we said we wanted to do with all of the `all.*` files…

colinrotherham commented:
Thanks @36degrees

>To avoid blocking this PR any further, shall we create an issue to look at which files/globs we include in export? (Before we ship v5)

Yeah sounds good. Other than ES module bundles + GitHub release JS, nothing that wasn't in the package before

>Just remove the version number, leaving `govuk-frontend.min.js`?

With the namespace directory maybe ~`govuk/govuk-frontend.min.js`~ `govuk/frontend.min.js`?

>I don't actually remember what we said we wanted to do with all of the `all.*` files…

We just had too many of them! Narrowed down a lot since then:

~`all.cjs`~
`all.mjs`
`all.bundle.js`
`all.bundle.mjs`
~`all.bundle.min.js`~
~`all.bundle.min.mjs`~ → `govuk-frontend.min.js`

36degrees commented:
> > Just remove the version number, leaving `govuk-frontend.min.js`?
> 
> With the namespace directory maaaybe ~`govuk/govuk-frontend.min.js`~ `govuk/frontend.min.js`??

I reckon `govuk-frontend.min.js` just to be really clear, if the file is copied or routed without keeping the directory name intact it'd still be clear what was being imported if you see it in a script tag.

> > I don't actually remember what we said we wanted to do with all of the `all.*` files…
> 
> We just had too many of them! Narrowed down a lot since then:
> 
> ~`all.cjs`~
> `all.mjs`
> `all.bundle.js`
> `all.bundle.mjs`
> ~`all.bundle.min.js`~
> ~`all.bundle.min.mjs`~ → `govuk-frontend.min.js`

Would it be more idiomatic to go with `index.*`? Otherwise I think we're good, does seem a lot more manageable with the other changes.

36degrees reviewed (approved):
Thanks for your patience as we worked through all of this @colinrotherham 🙌🏻

colinrotherham commented:
Ready for review again

We've removed the version suffix from the "browser ready" minified bundle:

```mjs
import { initAll } from '/javascripts/govuk-frontend.min.js'
```

I've updated our decision document to reflect this:

* https://github.com/alphagov/govuk-design-system-architecture/pull/28

colinrotherham commented:
We'll update our [`README.md` **Importing JavaScript**](https://github.com/alphagov/govuk-frontend/blob/main/packages/govuk-frontend/README.md#importing-javascript) advice in another PR:

>You can include Javascript for all components either by copying the `all.bundle.js` from `node_modules/govuk-frontend/dist/govuk/` into your application or referencing the file directly:
>
>```html
><script type="module" src="<path-to-govuk-frontend-all-file>/all.bundle.js"></script>
>```

☝️ It's still using the UMD + window globals approach but via `<script type="module">`
